File: dashboard/lib/permissions.ts
/**
 * Client-side permission helpers (UX only). Backend enforces authorization.
 * Strings must match core/pkg/authorization/permissions.go (RBAC v2).
 */
export const P = {
  authSession: 'auth.session',
  authPasswordChange: 'auth.password.change',
  authRegister: 'auth.register',

  findingsRead: 'findings.read',
  findingsAck: 'findings.ack',
  findingsDismiss: 'findings.dismiss',
  findingsResolve: 'findings.resolve',
  findingsReopen: 'findings.reopen',
  findingsBulk: 'findings.bulk',
  findingsDelete: 'findings.delete',
  findingsExceptionCreate: 'findings.exception.create',
  findingsExceptionApprove: 'findings.exception.approve',
  findingsExceptionDelete: 'findings.exception.delete',

  investigationsRead: 'investigations.read',
  investigationsWrite: 'investigations.write',
  investigationsDelete: 'investigations.delete',

  riskEvaluate: 'risk.evaluate',

  policiesRead: 'policies.read',
  policiesDraft: 'policies.draft',
  policiesReview: 'policies.review',
  policiesApprove: 'policies.approve',
  policiesPublish: 'policies.publish',
  policiesDelete: 'policies.delete',

  rulesRead: 'rules.read',
  rulesWrite: 'rules.write',
  rulesDelete: 'rules.delete',
  rulesImport: 'rules.import',
  rulesExport: 'rules.export',

  inventoryRead: 'inventory.read',
  inventoryAnnotate: 'inventory.annotate',
  inventoryModify: 'inventory.modify',
  inventoryQuarantine: 'inventory.quarantine',
  inventoryDelete: 'inventory.delete',
  inventoryBulk: 'inventory.bulk',

  runtimeRead: 'runtime.read',
  runtimeMappingWrite: 'runtime.mapping.write',

  graphReadSummary: 'graph.read.summary',
  graphReadPaths: 'graph.read.paths',
  graphQueryEntity: 'graph.query.entity',
  graphQueryTraversal: 'graph.query.traversal',
  graphQueryAdvanced: 'graph.query.advanced',
  graphExport: 'graph.export',

  /** Governed egress for risk findings export (CSV/PDF). */
  exportFindings: 'export.findings',

  /** JWT-bound server sessions (list/revoke). */
  sessionsRead: 'sessions.read',
  sessionsRevoke: 'sessions.revoke',
  sessionsRevokeAll: 'sessions.revoke_all',

  malwareRead: 'malware.read',
  malwareUpload: 'malware.upload',

  usersRead: 'users.read',
  usersCreate: 'users.create',
  usersUpdate: 'users.update',
  usersDisable: 'users.disable',
  usersDelete: 'users.delete',
  usersPasswordReset: 'users.password.reset',
  usersRoleAssign: 'users.role.assign',

  systemAuditRead: 'system.audit.read',
  observabilityMetricsRead: 'observability.metrics.read',
  observabilityLogsRead: 'observability.logs.read',
  observabilityAgentsRead: 'observability.agents.read',
  observabilityDebugRead: 'observability.debug.read',

  systemDebug: 'system.debug',
  internalCveTrigger: 'internal.cve.trigger',
  clusterCertificatesRotate: 'cluster.certificates.rotate',
} as const;

export type PermissionString = (typeof P)[keyof typeof P];

/** Build a Set of lowercase permission strings from user.permissions (if present). */
export function permissionSet(user: { permissions?: string[] } | null | undefined): Set<string>;
/** @returns A set containing all granted permissions in lowercase for case-insensitive comparison. */
export function permissionSet(user: { permissions?: string[] } | null | undefined): Set<string> {
  const list = user?.permissions ?? [];
  return new Set(list.map((s) => String(s).toLowerCase()));
}

/** Check if a single permission is granted to the user. */
export function can(user: { permissions?: string[] } | null | undefined, need: string): boolean;
/** @returns true if the user has the specified permission (case-insensitive). */
export function can(user: { permissions?: string[] } | null | undefined, need: string): boolean {
  return permissionSet(user).has(need.toLowerCase());
}

/** Check if any of the listed permissions are granted to the user. */
export function canAny(user: { permissions?: string[] } | null | undefined, needs: string[]): boolean;
/** @returns true if at least one permission in the list is granted (case-insensitive). */
export function canAny(user: { permissions?: string[] } | null | undefined, needs: string[]): boolean {
  const s = permissionSet(user);
  return needs.some((n) => s.has(n.toLowerCase()));
}

/** Check if all of the listed permissions are granted to the user. */
export function canAll(user: { permissions?: string[] } | null | undefined, needs: string[]): boolean;
/** @returns true only if every permission in the list is granted (case-insensitive). Useful for compound actions like bulk operations. */
export function canAll(user: { permissions?: string[] } | null | undefined, needs: string[]): boolean {
  const s = permissionSet(user);
  return needs.every((n) => s.has(n.toLowerCase()));
}

File: core/pkg/authorization/permissions.go
// Package authorization implements Fortuna application RBAC (permission-based).
// RBAC v2: fine-grained permissions (FORTUNA_RBAC_V2_ENHANCEMENT_SPEC). Deny-by-default.
// Distinct from core/pkg/rbac (Kubernetes RBAC analysis).
package authorization

// Permission is an atomic application capability string (stable API contract).
type Permission string

const (
	// Auth & session
	PermissionAuthSession        Permission = "auth.session"
	PermissionAuthPasswordChange Permission = "auth.password.change"
	PermissionAuthRegister       Permission = "auth.register"

	// Sessions (JWT binding + governance)
	PermissionSessionsRead       Permission = "sessions.read"
	PermissionSessionsRevoke     Permission = "sessions.revoke"
	PermissionSessionsRevokeAll  Permission = "sessions.revoke_all"

	// Export governance (sensitive data egress)
	PermissionExportFindings Permission = "export.findings"
	PermissionUsersRead         Permission = "users.read"
	PermissionUsersCreate       Permission = "users.create"
	PermissionUsersUpdate       Permission = "users.update"
	PermissionUsersDisable      Permission = "users.disable"
	PermissionUsersDelete       Permission = "users.delete"
	PermissionUsersPasswordReset Permission = "users.password.reset"
	PermissionUsersRoleAssign    Permission = "users.role.assign"

	// Findings (v2 — replaces findings.write / findings.exceptions)
	PermissionFindingsRead               Permission = "findings.read"
	PermissionFindingsAck              Permission = "findings.ack"
	PermissionFindingsDismiss          Permission = "findings.dismiss"
	PermissionFindingsResolve          Permission = "findings.resolve"
	PermissionFindingsReopen           Permission = "findings.reopen"
	PermissionFindingsBulk             Permission = "findings.bulk"
	PermissionFindingsDelete           Permission = "findings.delete"
	PermissionFindingsExceptionCreate  Permission = "findings.exception.create"
	PermissionFindingsExceptionApprove Permission = "findings.exception.approve"
	PermissionFindingsExceptionDelete  Permission = "findings.exception.delete"

	// Investigations (SOC case workspace)
	PermissionInvestigationsRead   Permission = "investigations.read"
	PermissionInvestigationsWrite  Permission = "investigations.write"
	PermissionInvestigationsDelete Permission = "investigations.delete"

	PermissionRiskEvaluate Permission = "risk.evaluate"

	// Policies (v2 — replaces policies.write)
	PermissionPoliciesRead    Permission = "policies.read"
	PermissionPoliciesDraft     Permission = "policies.draft"
	PermissionPoliciesReview    Permission = "policies.review"
	PermissionPoliciesApprove   Permission = "policies.approve"
	PermissionPoliciesPublish   Permission = "policies.publish"
	PermissionPoliciesDelete    Permission = "policies.delete"

	PermissionRulesRead   Permission = "rules.read"
	PermissionRulesWrite  Permission = "rules.write"
	PermissionRulesDelete Permission = "rules.delete"
	PermissionRulesImport Permission = "rules.import"
	PermissionRulesExport Permission = "rules.export"

	// Inventory (v2 — replaces inventory.write)
	PermissionInventoryRead       Permission = "inventory.read"
	PermissionInventoryAnnotate   Permission = "inventory.annotate"
	PermissionInventoryModify     Permission = "inventory.modify"
	PermissionInventoryQuarantine Permission = "inventory.quarantine"
	PermissionInventoryDelete     Permission = "inventory.delete"
	PermissionInventoryBulk       Permission = "inventory.bulk"

	PermissionRuntimeRead         Permission = "runtime.read"
	PermissionRuntimeMappingWrite Permission = "runtime.mapping.write"

	// Graph (v2 — replaces graph.read + graph.query.safe)
	PermissionGraphReadSummary   Permission = "graph.read.summary"
	PermissionGraphReadPaths     Permission = "graph.read.paths"
	PermissionGraphQueryEntity   Permission = "graph.query.entity"
	PermissionGraphQueryTraversal Permission = "graph.query.traversal"
	PermissionGraphQueryAdvanced Permission = "graph.query.advanced"
	PermissionGraphExport        Permission = "graph.export"

	PermissionMalwareRead   Permission = "malware.read"
	PermissionMalwareUpload Permission = "malware.upload"

	PermissionInternalCVETrigger Permission = "internal.cve.trigger"
	PermissionSystemDebug        Permission = "system.debug"
	PermissionSystemAuditRead    Permission = "system.audit.read"

	// Observability (v2 — replaces system.observability.read)
	PermissionObservabilityMetricsRead Permission = "observability.metrics.read"
	PermissionObservabilityLogsRead    Permission = "observability.logs.read"
	PermissionObservabilityAgentsRead  Permission = "observability.agents.read"
	PermissionObservabilityDebugRead   Permission = "observability.debug.read"

	PermissionClusterCertificatesRotate Permission = "cluster.certificates.rotate"
)

// PermissionLevel classifies sensitivity / SoD (spec §9).
type PermissionLevel string

const (
	LevelRead              PermissionLevel = "READ"
	LevelWrite             PermissionLevel = "WRITE"
	LevelDestructive       PermissionLevel = "DESTRUCTIVE"
	LevelPlatform          PermissionLevel = "PLATFORM"
	LevelSecurityCritical  PermissionLevel = "SECURITY_CRITICAL"
)

// ClassifyPermission returns coarse classification for governance UI / policy.
func ClassifyPermission(p Permission) PermissionLevel {
	switch p {
	case PermissionUsersDelete, PermissionUsersRoleAssign, PermissionUsersPasswordReset,
		PermissionSessionsRevokeAll,
		PermissionFindingsDelete, PermissionFindingsBulk,
		PermissionInvestigationsDelete,
		PermissionInventoryDelete, PermissionInventoryBulk, PermissionInventoryQuarantine,
		PermissionRulesDelete, PermissionPoliciesDelete,
		PermissionMalwareUpload, PermissionInternalCVETrigger,
		PermissionClusterCertificatesRotate:
		return LevelDestructive
	case PermissionUsersCreate, PermissionUsersUpdate, PermissionUsersDisable,
		PermissionSessionsRevoke,
		PermissionFindingsAck, PermissionFindingsDismiss, PermissionFindingsResolve, PermissionFindingsReopen,
		PermissionFindingsExceptionCreate, PermissionFindingsExceptionApprove, PermissionFindingsExceptionDelete,
		PermissionInvestigationsWrite,
		PermissionInventoryAnnotate, PermissionInventoryModify,
		PermissionRulesWrite, PermissionRulesImport,
		PermissionPoliciesDraft, PermissionPoliciesReview, PermissionPoliciesApprove, PermissionPoliciesPublish,
		PermissionRuntimeMappingWrite, PermissionRiskEvaluate,
		PermissionExportFindings:
		return LevelWrite
	case PermissionSystemDebug, PermissionObservabilityDebugRead:
		return LevelPlatform
	case PermissionAuthRegister, PermissionSystemAuditRead:
		return LevelSecurityCritical
	default:
		return LevelRead
	}
}

// AllPermissions returns every known permission (admin grant set).
func AllPermissions() []Permission {
	return []Permission{
		PermissionAuthSession,
		PermissionAuthPasswordChange,
		PermissionAuthRegister,
		PermissionSessionsRead,
		PermissionSessionsRevoke,
		PermissionSessionsRevokeAll,
		PermissionExportFindings,
		PermissionUsersRead,
		PermissionUsersCreate,
		PermissionUsersUpdate,
		PermissionUsersDisable,
		PermissionUsersDelete,
		PermissionUsersPasswordReset,
		PermissionUsersRoleAssign,
		PermissionFindingsRead,
		PermissionFindingsAck,
		PermissionFindingsDismiss,
		PermissionFindingsResolve,
		PermissionFindingsReopen,
		PermissionFindingsBulk,
		PermissionFindingsDelete,
		PermissionFindingsExceptionCreate,
		PermissionFindingsExceptionApprove,
		PermissionFindingsExceptionDelete,
		PermissionInvestigationsRead,
		PermissionInvestigationsWrite,
		PermissionInvestigationsDelete,
		PermissionRiskEvaluate,
		PermissionPoliciesRead,
		PermissionPoliciesDraft,
		PermissionPoliciesReview,
		PermissionPoliciesApprove,
		PermissionPoliciesPublish,
		PermissionPoliciesDelete,
		PermissionRulesRead,
		PermissionRulesWrite,
		PermissionRulesDelete,
		PermissionRulesImport,
		PermissionRulesExport,
		PermissionInventoryRead,
		PermissionInventoryAnnotate,
		PermissionInventoryModify,
		PermissionInventoryQuarantine,
		PermissionInventoryDelete,
		PermissionInventoryBulk,
		PermissionRuntimeRead,
		PermissionRuntimeMappingWrite,
		PermissionGraphReadSummary,
		PermissionGraphReadPaths,
		PermissionGraphQueryEntity,
		PermissionGraphQueryTraversal,
		PermissionGraphQueryAdvanced,
		PermissionGraphExport,
		PermissionMalwareRead,
		PermissionMalwareUpload,
		PermissionInternalCVETrigger,
		PermissionSystemDebug,
		PermissionSystemAuditRead,
		PermissionObservabilityMetricsRead,
		PermissionObservabilityLogsRead,
		PermissionObservabilityAgentsRead,
		PermissionObservabilityDebugRead,
		PermissionClusterCertificatesRotate,
	}
}

var validPermissionSet map[Permission]struct{}

func init() {
	validPermissionSet = make(map[Permission]struct{}, len(AllPermissions()))
	for _, p := range AllPermissions() {
		validPermissionSet[p] = struct{}{}
	}
}

// IsValidPermission is true only for permissions in the v2 taxonomy (unknown strings are invalid).
func IsValidPermission(p Permission) bool {
	_, ok := validPermissionSet[p]
	return ok
}
